game-core/src/main/java/games/strategy/engine/framework/startup/ui/SetupPanel.java

```diff
@@ -32,7 +32,9 @@
   private final List<Consumer<SetupPanel>> listeners = new CopyOnWriteArrayList<>();
 
   public void addObserver(final Consumer<SetupPanel> observer) {
-    listeners.add(observer);
+    if (!listeners.contains(observer)) {
```

Comment:
- I wonder if the observer list should actually be a set?
- Is there another fix out there so we don't try to add duplicate Observers?

Reply:
> I wonder if the observer list should actually be a set?

I considered changing it from CopyOnWriteArrayList to CopyOnWriteArraySet. But I think set iteration order is not defined and I didn't want to change any behaviour beyond this fix.

> Is there another fix out there so we don't try to add duplicate Observers?

Well, the calling code could ask this class if it's already an observer and not add itself, but that seems extra complexity for no reason.

Reply:
My guess/bet would be that the call order does not matter. Considering the caller spams 'addObserver', any ordering dependency is perhaps more accidental then intentional.

> Well, the calling code could ask this class if it's already an observer and not add itself, but that would just be extra complexity without a benefit.

Sounds reasonable.